### ISSUE TITLE
Action Recognition -- Separating the encoder and the decoder

### ISSUE THREAD

**Issue by bozic-djordje** [open-edge-platform/geti#103]
The Inference Engine (IE) [action recognition sample](https://docs.openvinotoolkit.org/latest/_inference_engine_ie_bridges_python_sample_action_recognition_README.html) takes  separated encoder and decoder pre-trained models as input arguments. On the other hand, when we export the Pytorch model using [the instructions provided](https://github.com/opencv/openvino_training_extensions/tree/develop/pytorch_toolkit/action_recognition), we get a single encoder-decoder architecture. I have two main questions:

1. Is there any way to export a Pytorch model into Openvino encoder model and Openvino decoder model separately so that they can be plugged into the [action recognition IE sample](https://docs.openvinotoolkit.org/latest/_inference_engine_ie_bridges_python_sample_action_recognition_README.html)?
2. Is there any way to train just the decoder part of the architecture?

**Comment by snosov1** [comment]
@vadimadr can you, please, comment?

**Comment by vadimadr** [comment]
Hi, @Dolokhow! 
1. You can export encoder and decoder models separately. To do this you need to provide *-vtn_encoder or *-vtn_decoder in `--model`. Note that if you use IE sample for inference, you must provide normalization and reverse channel order options during conversion of encoder in `mo.py`, decoder model does not require any pre-processing.  
2. There is no option to train encoder-only or decoder-only, but you can set zero learning rate to the parameters that you want to freeze. To do this you need to insert regex for parameter names you want to freeze in `trainable_parametrs()` method of corresponding model. This, however, won't save you any memory usage (pytorch still saves gradients). Hence if you want to fine-tune decoder and save some GPU memory you can try to wrap non trainable parts of the architecture in `with torch.no_grad()`

**Comment by Dolokhow** [comment]
Hi @vadimadr,

Thank you for a very quick response. I've just tested what you have suggested and I have just two more questions before I close this issue. I'll also be adding a couple of comments and clarifications so that others can repeat this process more easily, please correct me if I am wrong somewhere.

Questions:

1. For conversion from ONNX -> VINO format of the encoder, you've said that one needs to provide normalization and reverse channel order options. The reverse channel option is --reverse_input_channels, can you give me the explicit flag name for the normalization option?
2. [Read after reading Clarifications section below]: If one wanted to convert the entire architecture (encoder+decoder) into a single VINO .xml + .bin pair, given the fact that --reverse_input_channels option does not work, how should one solve that situation? Specifically, if I understand correctly, if channels are not reversed then that would have significant impact on the accuracy of the model, and if there is no way to apply this operation on the entire ONNX architecture (encoder+decoder) then the only way to use the model is to separate it into the encdoer part and the decoder part?

Clarifications:

1.  Using --reverse_input_channels will not work when trying to convert the entire architecture (encoder+decoder) defined in a single ONNX file to VINO format. The error raised is below: 
 > [ ERROR ]  Reverse input channels are not applied -- appropriate convolutions were not found
2. The --input_shape for the entire (encoder+decoder) architecture when coverting from ONNX to VINO is '[1,16,3,224,224]' as is given in the [instructions](https://github.com/opencv/openvino_training_extensions/tree/develop/pytorch_toolkit/action_recognition). However, when converting just the extracted ONNX encoder, the --input_shape is [1,3,224,224].
3. The --input_shape when converting the ONNX decoder to VINO format is [1,16,512] if you use --clip-size 16 during training.

**Comment by vadimadr** [comment]
1. The options for normalization are: `--mean_value` and `--scale_values`. @AlexanderDokuchaev could you please provide the correct line with exact values? 
2. You are correct about impact of incorrect pre-processing on model accuracy.  You can always skip the pre-processing  options in `mo.py` and rely on normalization in your sample code. In this case, you need to implement the same operations that you used to prepare batch during the training in your code (e.g. subtract men, divide by scaling factor, convert color to BGR). But I hope the bug with pre-processing of 3D inputs will be fixed soon.

**Comment by Dolokhow** [comment]
@AlexanderDokuchaev Could you please give me the exact values, or tell me where to look for them?

**Comment by AlexanderDokuchaev** [comment]
Sorry, `--mean_values=[103.53,116.28,123.675] --scale_values=[58.395,57.12,57.37499999999999]`

**Comment by Dolokhow** [comment]
Thank you all for your help! I am closing the issue.
